Fixing dialog closing unexpectedly when typing in the textarea

Rishab87 opened the pull request:
<!--
Thanks for contributing to JupyterLab!
Please fill out the following items to submit a pull request.
See the contributing guidelines for more information:
https://github.com/jupyterlab/jupyterlab/blob/main/CONTRIBUTING.md
-->

## References

Fixes #14511 
<!-- Note issue numbers this pull request addresses (should be at least one, see contributing guidelines above). -->

<!-- Note any other pull requests that address this issue and how this pull request is different. -->

## Code changes
I've added test in my first commit as told by @krassowski in discussions of this issue. This test is failing which is the expected behaviour for now:

```bash
 ● @jupyterlab/apputils › Dialog › #handleEvent() › keydown › should not accept on enter key within textarea

    expect(received).toBe(expected) // Object.is equality

    Expected: true
    Received: false

      213 |           }
      214 |   
    > 215 |           expect(dialog.isVisible).toBe(true);
          |                                    ^
      216 |   
      217 |           simulate(dialog.node, 'keydown', { key: 'Enter', keyCode: 13 });
      218 |   

      at Object.<anonymous> (test/dialog.spec.tsx:215:36)


 RUNS  test/sessioncontext.spec.ts
```

<!-- Describe the code changes and how they address the issue. -->

## User-facing changes

<!-- Describe any visual or user interaction changes and how they address the issue. -->

<!-- For visual changes, include before and after screenshots or GIF/mp4/other video demo here. -->

## Backwards-incompatible changes

<!-- Describe any backwards-incompatible changes to JupyterLab public APIs. -->

krassowski commented:
The test looks good to me!

krassowski commented on packages/apputils/src/dialog.tsx:
```diff
@@ -440,8 +440,8 @@ export class Dialog<T> extends Widget {
 
         if (activeEl instanceof HTMLButtonElement) {
           index = this._buttonNodes.indexOf(activeEl);
```
Can we now defined index as `const` here and remove line 439?

krassowski commented:
It looks like test failures are related:

```
 FAIL  test/dialog.spec.tsx
  ● @jupyterlab/apputils › Dialog › #handleEvent() › keydown › should resolve with currently focused button

    thrown: undefined

      at _getError (../../node_modules/jest-circus/build/utils.js:424:18)
          at Array.map (<anonymous>)
```

and a number of tests timing out, e.g.:

```
  ● @jupyterlab/apputils › InputDialog › getBoolean() › should be false by default

    thrown: "Exceeded timeout of 10000 ms for a test.
```

I think that these tests may need work.

Rishab87 commented:
@krassowski, I've fixed tests timing out issue but but for thrown undefined I've observed few things:
- index was coming undefined in one case but then I used default value if index not present but still getting that thrown undefined error.
- after           simulate(dialog.node, 'keydown', { keyCode: 13 }); in should resolve with currently focused button test case I'm getting document active element as htmlBodyElement but inside main file I was getting htmlButtonElement as active for this test case.
 here is my current implementation:
 ```javascript
  case 13: {
        // Enter.
        event.stopPropagation();
        event.preventDefault();

        const activeEl = document.activeElement;
        let index:number | undefined = this._defaultButton;


        if (activeEl instanceof HTMLButtonElement) {
          index = this._buttonNodes.indexOf(activeEl);
          this.resolve(index !== -1 ? index : 0);
        } else if(!(activeEl instanceof HTMLTextAreaElement) || !activeEl){
          this.resolve(index !== undefined ? index : 0);        
        }
        break;
      }
```
I've tried many things but not able to resolve this test case, can you please guide me on this?

krassowski commented:
Not a good idea immediately - if you could push your changes I can try to poke at it locally later

Rishab87 commented:
@krassowski , I've pushed the code

Rishab87 commented:
@krassowski , I've fixed the issue now all tests pass, the issue was because every test case needs to be resolved in jest otherwise it will throw undefined but if we resolve in case of this issue then dialog box will close which is not the expected behaviour so promise should not be resolved in this case so I added a generalized work around for this which can help in future test cases too where promises should not resolve.

Rishab87 commented:
@krassowski anything else I need to change here?

krassowski commented on packages/apputils/test/dialog.spec.tsx:
```diff
@@ -178,6 +223,35 @@ describe('@jupyterlab/apputils', () => {
+
+          const promptPromise = dialog.launch();
+
+          await dialog.resolveFor(async () => {
+            const textarea = dialog.node.querySelector('textarea');
+            textarea!.focus();
+
+            simulate(dialog.node, 'keydown', { keyCode: 13 });
+
+            await new Promise(resolve => setTimeout(resolve, 50));
+          });
```
If you undo your changes in `dialog.tsx`, does it pass, or does it fail?

krassowski replied:
I just struggle to follow this test, it feels like we need to simplify the logic.

Rishab87 replied:
You're right, I've fixed it now and implemented a much simplified approach, I should've thought it carefully before pushing, now it works as expected

krassowski commented on packages/apputils/src/dialog.tsx:
```diff
@@ -436,12 +436,14 @@ export class Dialog<T> extends Widget {
         if (activeEl instanceof HTMLButtonElement) {
           index = this._buttonNodes.indexOf(activeEl);
+          this.resolve(index !== -1 ? index : 0);
+        } else if (!(activeEl instanceof HTMLTextAreaElement) || !activeEl) {
```
It is not immediately clear to me why there is `|| !activeEl` part in here. Would it still work without it?

Rishab87 replied:
Yes it still works without it, made the changes

krassowski commented on packages/apputils/src/dialog.tsx:
```diff
@@ -436,12 +436,14 @@ export class Dialog<T> extends Widget {
 
         if (activeEl instanceof HTMLButtonElement) {
           index = this._buttonNodes.indexOf(activeEl);
+          this.resolve(index !== -1 ? index : 0);
```
Thinking about why index might be `-1`. This could be if the react body had a button which is not a button node. I think in that case we would not want to resolve either. Say if there was "Help" button in extension author's code.

krassowski replied:
I will try that and see what tests on CI think:

```suggestion
          if (index !== -1) {
            this.resolve(index);
          }
```

krassowski commented on packages/apputils/src/dialog.tsx:
```diff
@@ -436,12 +436,14 @@ export class Dialog<T> extends Widget {
+        let index: number | undefined = this._defaultButton;
 
         if (activeEl instanceof HTMLButtonElement) {
           index = this._buttonNodes.indexOf(activeEl);
```
Index here is always overwritten, we may as well remove the assignment above to reduce cognitive complexity

```suggestion
          const index = this._buttonNodes.indexOf(activeEl);
```

krassowski commented on packages/apputils/src/dialog.tsx:
```diff
@@ -436,12 +436,14 @@ export class Dialog<T> extends Widget {
 
         const activeEl = document.activeElement;
-        let index: number | undefined;
+        let index: number | undefined = this._defaultButton;
```
```suggestion
```

krassowski commented on packages/apputils/src/dialog.tsx:
```diff
@@ -436,12 +436,14 @@ export class Dialog<T> extends Widget {
           index = this._buttonNodes.indexOf(activeEl);
+          this.resolve(index !== -1 ? index : 0);
+        } else if (!(activeEl instanceof HTMLTextAreaElement)) {
+          this.resolve(index !== undefined ? index : 0);
```
Index here is always `this._defaultButton` which is never undefined. Against, we can simplify the code:

```suggestion
          const index = this._defaultButton;
          this.resolve(index);
```

krassowski reviewed (approved):
Thank you @Rishab87!

krassowski commented:
@meeseeksdev please backport to 4.3.x
